rfd/0193-access-control-decision-api.md

```diff
@@ -0,0 +1,263 @@
+
+During this phase we will also implement tctl commands for directly invoking the DecisionService gRCP API on the control
+plane. This will provide a powerful debugging and auditing tool for superusers and developers who want deeper insight into how
+teleport makes decisions.
+
+#### Relocate Phase
+
+The relocate phase will see us actually transition to all access-control decisions being made at the control plane. The method
+of relocation will depend on the kind of teleport agent. For tunnel agents, we will start performing denial at the control plane
+itself, and allowed dials will have the Permit object forwarded to the agent as part of the dial. For direct dial agents, the
+agent will call-back into the control plane to get a decision.
+
+We intend to open a second follow-up RFD when we are closer to the relocate phase in order to explore it in more detail, but the
+highlights are these:
+
+- Agent reverse tunnel protocols will need to be updated to allow a trusted and replay-resistant permit message to be sent from proxy to
+agent as part of an incoming dial.
+
+- The mechanism by which trusted cluster dials are performed will need to be reworked s.t. all routing decisions are made at
+the leaf proxy rather than the root proxy since access-control decisions will now get made as a part of routing. This will include
+sending sideband information about user identity when forwarding the dial from root proxy to leaf proxy in order to ensure that
+the leaf proxy can correctly map the identity without relying on user certificate information.
+
+- One major version of buffer will be needed during which agents still cache roles locally and roles/traits are still applied to certs
+so that we can fallback to local decisions when talking to outdated control plane elements.
+
+- Agentless/openssh nodes are already using a "decisions at control plane" model, but they still depend on logins being present in
+certificates. We may want to consider a mechanism for eliminating logins at the same time (e.g. by lazily provisioning certs containing
+only the target login at the proxy).
+
+### Questions/Alternatives/Ongoing Research
+
+- We've gone back and forth a bit on whether to return bare permits s.t. the resulting client method returns `(*Permit, error)` or
+to use an enclosing response/decision message with a oneof for allow/deny cases. The former is more inline with how existing access
+control decisions are made and results in slightly cleaner code, but the latter may be useful if we ever decide to start doing more
+complex denial cases (e.g. for MFA flows or verbose denials for use by visualization tools). We're currently leaning toward bare
+permits since all the good arguments for enclosed responses are speculative at this point. One of the design goals of this system
+is to let us iterate faster, and that includes adding new methods with new conventions as the need arises.
```

Comment:
It'd be work, but nothing crazy. The idea is that this would function just like changing the return type of any other GRPC API... add an `EvaluateSSHAccessV2` (or create a v2 of the service as a whole if you're changing all of them), and continue to also support the old method for a major version or two.  Since control-flow already splits depending on approval/denial, most of your happy path continues to use the `Permit` object like before so there isn't a lot of downstream code churn past the decision call location.

Reply:
Based on discussions here and elsewhere, we're switching to "wrapped" decision responses from the get-go.